src/Native/Runtime/unix/UnixEvent.cpp

```diff
@@ -0,0 +1,334 @@
+
+    st = pthread_condattr_destroy(&attrs);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter condition attribute");
+}
+
+UnixEventWaiter::~UnixEventWaiter()
+{
+    int st = pthread_mutex_destroy(&m_mutex);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter mutex");
+
+    st = pthread_cond_destroy(&m_condition);
+    FATAL_ASSERT(st == 0, "Failed to destroy UnixEventWaiter condition variable");
+
+    if (m_waiterListEntries != m_embeddedWaiterListEntries)
+    {
+        free(m_waiterListEntries);
+    }
+}
+
+// Add event to the waiter
+// Parameters:
+//  index - Index of the event that the waiter reports when the event is signalled
+//  event - Event to wait on
+void UnixEventWaiter::AddEvent(int index, UnixEvent* event)
+{
+    ASSERT(index < m_eventCount);
+    m_waiterListEntries[index] = ListEntry(this);
+    event->AddWaiter(&m_waiterListEntries[index]);
+}
+
+// Remove event from the waiter
+// Parameters:
+//  index - Index of the event to remove
+//  event - Event to remove
+void UnixEventWaiter::RemoveEvent(int index, UnixEvent* event)
+{
+    ASSERT(index < m_eventCount);
+    event->RemoveWaiter(&m_waiterListEntries[index]);
+}
+
+// Signal the waiter that an attached event was set. It is called 
+// by the UnixEvent instances. 
+// Parameters:
+//  index - index of the event in the list of events managed by this waiter
+// Return:
+//  true if the signal has caused the wait to complete, false if the wait
+//  completion was already triggered by another signal.
+bool UnixEventWaiter::Signal(int index)
+{
+	bool releasedWait = false;
+    pthread_mutex_lock(&m_mutex);
+    // Only the first signal is used to release the waiting thread. 
+    // If the wait has already timed out or failed, we don't consume the signal.
+    // This is important for autoreset events
+    if (!m_timedOutOrFailed && (m_eventIndex == NoEventIndex))
+    {
+	    m_eventIndex = index;
+	    releasedWait = true;
+	    // Unblock the thread waiting for the condition variable
+	    pthread_cond_signal(&m_condition);
+	}
+    pthread_mutex_unlock(&m_mutex);
+
+    return releasedWait;
+}
+
+// Wait for one of the events attached to this waiter. It completes when
+// either one of the events is set or the wait timeouts.
+// Parameters:
+//  milliseconds - wait timeout
+// Return:
+//  One of the following values:
+//      WAIT_OBJECT_0 + n - the wait completed due to the event with index n
+//      WAIT_TIMEOUT      - the wait timed out
+//      WAIT_FAILED       - the wait has failed due to some system related issue
+uint32_t UnixEventWaiter::Wait(uint32_t milliseconds)
+{
+    timespec endTime;
+#ifdef __APPLE__
+    uint64_t endMachTime;
+    mach_timebase_info_data_t timeBaseInfo;
+#endif
+
+    if (milliseconds != INFINITE)
+    {
+#if HAVE_CLOCK_MONOTONIC
+        clock_gettime(CLOCK_MONOTONIC, &endTime);
+        TimeSpecAdd(&endTime, milliseconds);
+#else // HAVE_CLOCK_MONOTONIC
+
+#ifdef __APPLE__
```

Comment:
I prefer the `__APPLE__`. The mach absolute time is an concept that exists for Apple only.